### DISCUSSION TITLE
Support for arm64

### DISCUSSION THREAD

**Discussion by AsrieltheGoat** [danny-avila/LibreChat#1135] [Troubleshooting]
### Contact Details

asriel@dreemurrs.eu.org

### What features would you like to see added?

Can you guys add support for pre-built docker images?
It can be really helpful for me and maybe others

### More details

I have compiled my software on the arm server and created Docker images that work perfectly on this platform. However, when I try to run the pre-built image from this repo, it doesn't seem to be compiled for arm64

### Pictures

![image](https://github.com/danny-avila/LibreChat/assets/25924894/ba79f3a4-3a3e-4eab-9b31-33e10f9721b7)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
I've not heard of this being an issue, and I'm not sure how I could make it compatible.

Which image are you using? https://github.com/danny-avila?tab=packages&repo_name=LibreChat

Also, From the docs: https://docs.librechat.ai/install/docker_compose_install.html#running-librechat

> Note: MongoDB does not support older ARM CPUs like those found in Raspberry Pis. However, you can make it work by setting MongoDB’s version to mongo:4.4.18 in docker-compose.yml, the most recent version compatible with

Have you done this?

**Reply by danny-avila** [reply]
The repo images are just node-based images so they should be compatible. You may be running into an issue with mongodb @ latest

**Reply by AsrieltheGoat** [reply]
Im using the main one: danny-avila/librechat:latest
I got it to run with a bit of tweaking to the docker-compose file

```version: "3.9"

services:
  # client:
  #   image: nginx-client
  #   build:
  #     context: .
  #     target: nginx-client
  #   restart: always
  #   ports:
  #     - 3080:80
  #   volumes:
  #     - /client/node_modules
  #   depends_on:
  #     - api
  api:
    container_name: LibreChat
    ports:
      - 3004:3080               # Change it to 9000:3080 to use nginx
    depends_on:
      - mongodb
    image: librechat                # Comment this & uncomment below to build from docker hub image
    build:                                   # ^------
      context: .                              # ^------
      target: node                             # ^------v
    # image: ghcr.io/danny-avila/librechat:latest # Uncomment this & comment above to build from docker hub image
    restart: always
    user: "${UID}:${GID}"
    extra_hosts: # if you are running APIs on docker you need access to, you will need to uncomment this line and next
    - "host.docker.internal:host-gateway"
    env_file:
      - .env
    environment:
      - HOST=0.0.0.0
      - MONGO_URI=mongodb://mongodb:27017/LibreChat
      # - CHATGPT_REVERSE_PROXY=http://host.docker.internal:8080/api/conversation # if you are hosting your own chatgpt reverse proxy with docker
      # - OPENAI_REVERSE_PROXY=http://host.docker.internal:8070/v1/chat/completions # if you are hosting your own chatgpt reverse proxy with docker
      - MEILI_HOST=http://meilisearch:7700
      - MEILI_HTTP_ADDR=meilisearch:7700
    volumes:
      - /app/client/node_modules
      - ./api:/app/api
      - ./.env:/app/.env
      - ./.env.development:/app/.env.development
      - ./.env.production:/app/.env.production
      - /app/api/node_modules
      - ./images:/app/client/public/images

  mongodb:
    container_name: chat-mongodb
    # ports:  # Uncomment this to access mongodb from outside docker, not safe in deployment
    #   - 27018:27017
    image: arm64v8/mongo:latest
    restart: always
    #user: "${UID}:${GID}"
    volumes:
      - ./data-node:/data/db
    command: mongod --noauth

  meilisearch:
    container_name: chat-meilisearch
    image: getmeili/meilisearch:latest
    restart: always
    # ports: # Uncomment this to access meilisearch from outside docker
    #   - 7700:7700 # if exposing these ports, make sure your master key is not the default value
    env_file:
      - .env
    #user: "${UID}:${GID}"
    environment:
      - MEILI_HOST=http://meilisearch:7700
      - MEILI_HTTP_ADDR=meilisearch:7700
      - MEILI_NO_ANALYTICS=true
    volumes:
      - ./meili_data:/meili_data
```
      
   
![image](https://github.com/danny-avila/LibreChat/assets/25924894/7b2b86e6-74bc-4a71-b203-33d92c3b4391)

**Reply by Nindaleth** [reply]
I'm running a RPi4 which is too old ARMv8 for Mongo - using the recommended older version instead and it works. However, LibreChat container itself still won't run:

```
# docker container run -it ghcr.io/danny-avila/librechat:latest --entrypoint /bin/bash
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
exec /usr/local/bin/docker-entrypoint.sh: exec format error
```

There's no way for me to download an appropriate image arch:
```
# docker image pull ghcr.io/danny-avila/librechat:latest --platform linux/arm64/v8
latest: Pulling from danny-avila/librechat
Digest: sha256:9eb63673822ee10ae2c69af157653b5166f6e419261950fb7284d31a49fff729
Status: Image is up to date for ghcr.io/danny-avila/librechat:latest
image with reference ghcr.io/danny-avila/librechat:latest was found but does not match the specified platform: wanted linux/arm64/v8, actual: linux/amd64
```

When building my own image, I'm having some issues too, but maybe that would deserve another discussion thread later without derailing this topic.

**Reply by danny-avila** [reply]
@Nindaleth I think a PR resolved this issue but it won't build another compatible image until the next tag. The good news is I will tag a release soon

**Reply by danny-avila** [reply]
for reference: https://github.com/danny-avila/LibreChat/pull/1607

**Reply by Nindaleth** [reply]
Oh, excellent! That's great to hear, I'll be looking forward to trying out the new release.

**Reply by Nindaleth** [reply]
I can confirm that starting with the release v0.6.9 there are images provided that work on RPi4, I could successfully deploy LibreChat and use it to have a quick chat via an API provider. Thanks a lot and happy birthday to LibreChat!

**Reply by danny-avila** [reply]
Awesome thanks for confirming!

**Reply by clarity99** [reply]
hmm, I get : "! api The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s" docker compose up on arm64 on oracle cloud arm servers

**Reply by danny-avila** [reply]
@clarity99 are you using the dev images or the release images? currently only the release images have arm64/amd64 support, but I can see about changing that.

You can also try building it yourself. This recently changed in the compose file to pull down the dev images instead of building it manually

**Reply by clarity99** [reply]
First it was dev image, as that was in the docker compose I got from the git pull command. I've also tried without dev, with latest and it's also missing arm64 build, but if I fix it at v0.6.9 it works. 
i tried building it, but that didn't work either, as build section seems to be missing from the docker-compose file and I didn't have time yet to look at it further.

**Reply by danny-avila** [reply]
working on adding arm64 support to the dev images

**Reply by clarity99** [reply]
thank you! in the meantime I also managed to build it manually, so I have the latest version running on arm as well.

**Reply by danny-avila** [reply]
nice! I successfully built the dev images in both amd64 and arm64 this morning so the latest tags should work for you now.

**Reply by lazka** [reply]
I can confirm that the arm64 dev images work nicely here. Thank you!
